mlflow/pyfunc/__init__.py

```diff
@@ -999,11 +999,18 @@ def _parse_spark_datatype(datatype: str):
     from pyspark.sql.functions import udf
     from pyspark.sql.session import SparkSession
 
-    returnType = "boolean" if datatype == "bool" else datatype
-    schema = (
-        SparkSession.active().range(0).select(udf(lambda x: x, returnType=returnType)("id")).schema
-    )
-    return schema[0].dataType
+    return_type = "boolean" if datatype == "bool" else datatype
+    try:
+        # SparkSession.active only exists for spark >= 3.5.0
+        schema = (
+            SparkSession.active()
+            .range(0)
+            .select(udf(lambda x: x, returnType=return_type)("id"))
+            .schema
+        )
+        return schema[0].dataType
+    except AttributeError:
+        return udf(lambda x: x, returnType=return_type).returnType
```

Comment:
`udf(lambda x: x, returnType=return_type).returnType` should be incompatible with spark connect mode

Reply:
Yes, so we default to this when spark < 3.5

Reply:
But spark 4.0 actually works without this change

Reply:
then I think we can write:

```
return_type = udf(lambda x: x, returnType=return_type).returnType

if isinstance(return_type, UnparsedDataType):
        schema = (
            SparkSession.active()
            .range(0)
            .select(udf(lambda x: x, returnType=return_type)("id"))
            .schema
        )
        return schema[0].dataType

return return_type
```

Reply:
`udf(lambda x: x, returnType=return_type).returnType` is more performant, we can try it firstly, if it does not work, fallback to `select(udf(lambda x: x, returnType=return_type)`

Reply:
UnparsedDataType only exists in spark 3.5 & above, we need to try catch the importError as well.

Reply:
Updated!